### PR TITLE
Fix the test for nodes

### LINKED ISSUE #8: Incorrect hashes in "nodes.json" + incorrect check
Hello, I was integrating this into TezEdge's tests, and for some reason the "nodes" tests was failing.

I noticed that the `check_file` function, always loads "inodes.json". The fix is:

```diff
diff --git a/bin/main.ml b/bin/main.ml
index 6ed3705..61b8de3 100644
--- a/bin/main.ml
+++ b/bin/main.ml
@@ -176,7 +176,7 @@ module Nodes = struct
     generate_file config Gen.long_inode Inodes
 
   let check_file dir f =
-    let path = file dir None Inodes in
+    let path = file dir None f in
     let msg =
       match f with Nodes -> "nodes" | Inodes -> "inodes" | _ -> assert false
     in
```

After changing this, I not see that the tests fail when running the OCaml code in this repo too (which now produces the same hashes as my Rust code).

The other problem (I think) is that the generator always uses `Inter.Val.v` which I guess is for inodes, and not `Nodes.v` when producing nodes. Is my understanding here correct or is there something else I am missing?

### PR DESCRIPTION
**Pull request by samoht**
Fixes #8

Thanks to @tizoc for the report.

### CODE REVIEW

**Comment by tizoc** [comment]
This is generating the same `nodes.json` file, so what changed was the check to verify it (which now uses the wrong hashes), but it is the hashes in the generated file that need to change, right?

File: bin/main.ml [hunk 1]
```diff
@@ -202,8 +198,8 @@ module Nodes = struct
                  { Spec.name; kind; hash })
         in
         let vs = List.map Spec.value bindings in
-        let n = Store.Private.Node.Val.v vs in
-        let h = Spec.hash_node n in
+        let n = Inter.Val.v vs in
+        let h = Inter.Val.hash n in
```
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply bin/main.ml:202]: But I'm unsure how this relates to the tests in the CI

**Review by Ngoguey42 — APPROVED** [review]

**Comment by samoht** [comment]
Do the hashes in `nodes.json` fail for the rust node?

File: bin/main.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (samoht) [reply bin/main.ml:202]: well `Store.Private.Node.Val.v |> Spec.hash_node` was failing only for short nodes!

**Comment by tizoc** [comment]
@samoht yes, testing against `inodes.json` works (uses the hashing function for >256 directory entries), but for `nodes.json` (using the old function) it fails:

```
Expected hash CoWCVQQRxmndUyNvCRpSn36viaGwp8MEJarqT3Xmkju5Apmj3WeK but got CoV8YLjMcZvaMFwddecEaXvcFjeFsak89PFjmc5SsNftPGx24u4L (bindings: 177)
Expected hash CoW1L4gKZsKib8ggRhURWMSSj8VAxewuqEMwqBYxF4CDzQmSZL3i but got CoVeRup6TjxzoacBJRFEtZE2fqnMkg38kgF5VN3QtJBRGTcq5rHm (bindings: 226)
Expected hash CoW6oe7AUF2Nan5P5MjGZ8vvACjoVxpg7drJQ6wAUzSMZmvFqLfp but got CoWBXS2huwhy2fRpNLQ1gFkoKCXqr2o2poGzX6oEioj3LBXq23uD (bindings: 33)
Expected hash CoWE85BL5KFVPowD2XQhNGoWz2sZwiEsg3o6FTpUxQfxLUusn4F7 but got CoWLnvLeuvYBFtwgFY1moxfPCAFU61MqPtTTtj5pf8bfCghMDKmx (bindings: 211)
Expected hash CoVzhLVMm8wND4dYyKjy2etGzy73RSML7Yn6YEbiA4WsXfiR8Nvp but got CoV7YXfbncChGdGfhCo4dP1Nds5JJHKATodveKoSfyLhXxf4eFqh (bindings: 12)
Expected hash CoWTbrhf7wuNEyf19VNKGHMkdm2GGr1qcqn2yit32v5J8MTBbxag but got CoVjytZUrp5YnhhNegUjHjFdkygRchkjGgEtTUznxN5dSFJzQUEg (bindings: 44)
Expected hash CoUrrbUpYRBd5ia6usTH2pr7YPfLWZKT5qVCwPYLNma2UfmHugPC but got CoWNwcazmXHoABaAyKaYUQNforvFcRFahxBTh2bngr5QCieN4ixZ (bindings: 109)
Expected hash CoVm4wVkztX9iS8jUn1tNN9xzXBJ3ypCP4B2GpBfXxZJKWTyJQsZ but got CoUrTjsWHsNFDsrm1rzZEc9XzR8deAozMbaDYqxPTiCRrUDp1M6z (bindings: 232)
Expected hash CoVnnPR2kL9nAZCFihEaWaVMaA4mXXvotZus1AztB8fXsakGc8PV but got CoVSYeSRLQy96JE7iH4skPGXaP1rHniCLhyo3fBtc6r7pdannwcq (bindings: 62)
Expected hash CoVpRu7vwS55rgDdnqoWyoANypfJsYowKRx2QG1fYTyecJPHgQGL but got CoVtPbdci7MWKUuvM8YfoJjNCmwRGZSrpaR2ztxYfndiJCX4wq7a (bindings: 194)
Expected hash CoWQnp82ea68Z2HyNJnSTfxX9WFet5sc7DKBPqrRtB6uV9jFvqpr but got CoVEyLPxdo9DmvaQySqGFsCLhru5x5G5fW6S6BxiUtTd2YMQkmM2 (bindings: 85)
Expected hash CoWD248E19QUW81o7GQ683VpN1U1HCtedgNggqpZD5adt1fP2swi but got CoVrLpmqGMCtwQ3h6uZ5J6Lpfj9ygSbdzx3cu18bZz6KxYweSC3b (bindings: 9)
Expected hash CoVFneVvvp3BFRqeguxQ5qWmhboKHkdkqX6difbdqudR4g8vEKrX but got CoUnvXm4V8fZph8oUiwLXBW6fTzGgagLzwQ6zLadioPz4VXYbmwi (bindings: 196)
Expected hash CoUtvZsPZunHrjk8SPbkn8r4yJgwsBGDY1JxKAbh4AYTPmxJHFnR but got CoVwXXVRrX5knXePB1uCS5faMRPmGwvt2D81AKLmmgiEgXVUeumr (bindings: 28)
Expected hash CoWRngih5u1NFwm1uZhJcwC4iNQiW7Ydp5hePdD3YiV1ypQfn76X but got CoUkSSPdRQf6PQWtfVdmd7kkGcJUC1wZihYBvmWjpdfkcbsxeJEV (bindings: 206)
Expected hash CoVKfttLQmt28kx9YHEbqTapxBCT9RtjtJLhjoLaK1t97H2b8Ws6 but got CoW2epQ9CHFLwJiYZHj8a8urWdoBbDqXShz1H6W4Cw3pjZqe5vBS (bindings: 24)
Expected hash CoVjtzB5jYF7EvXTxzLJc7mhfo47Je6asypD1C3qhfFGJwKVnpns but got CoVbBmWANP8Q2tLf1B2Ag5imKvMjkTo8tYLgoA4td6QZq2HjSyUE (bindings: 179)
Expected hash CoUknKP9oK3rBzSWkAc9ByZ5xcgAN5Wit4RujpzhJPHvv77LfwbF but got CoUmUaTT67iCPg3XgMnDKvvv6UA2naPBGLKX7i3RY4aQ9Z1zkTXa (bindings: 71)
Expected hash CoUreWV9nbVab8WX9jQZq59UPX6VraG64oQ1B3zKavfr7pToum5y but got CoVoQDGzhzLuw4xX5sVCobMH8ND1Fdu1DtSXhmB1mXgEUezu7pMc (bindings: 169)
Expected hash CoUrjKxSzA6cqMRkPQVXJcuZw1NpfRnu1m4u5yDaGkAf8rLo6AtF but got CoUv8Puyioseh3qFtrccUyvtAbARrnHz4KwCGb73zwvnsuNPKf4g (bindings: 7)
Expected hash CoVbknn46SBESrNJJCARKaqX3Bt5nWHZf1EfdsefupqMnVy11M3p but got CoWZ5tzKSAxoPdv5D6EY89k4WXyWY9uPgdHsDF8KBzuZM3ALQ42r (bindings: 219)
Expected hash CoV3EoWoCCk5zHHjCL4z2MdtbTZQrTFQuqg1RTqrTV4yRACJKQCq but got CoVPPrGvafXzBn4xLwVM8yvEsmdeNCk1Ch86kRuBJCpTbNbXLYww (bindings: 247)
Expected hash CoVYtaSCzzTnacRrGWRWMjpCMuGbqUUx54R4xgEsMSr3g7vgE2ML but got CoW4QbhmiAFUHqGK9McanGLtTGDfRUnzPD9HeENc4qL1K4CkUjVW (bindings: 212)
Expected hash CoWSKKuzAn7jc9QTCCkzhmTTpfA1aCnfmzoXLRgUiUY5q4Bv8rUN but got CoW2gePdD9eVdmYXdpyK2BCKoVsNHrdYezPysufSU8yVdajvL2hN (bindings: 134)
Expected hash CoVfU7LHz8FJhAJ1qHiKhm1jB1ZuNaGznvWfNn1asLWDoQV8PANd but got CoUxthHqYeSmvUGkrygcm2GfCgzgKsTxoeK8PwGMd5Qofqe8RotD (bindings: 237)
Expected hash CoVA739v4uC5fRAeDWwfrppm1CArccJ89U9Qgro8efhzVooBoyRR but got CoVPG8gbCJQxUCmuNZTLSys11aej8TCKmJZuwbeuzdz2aXu79nvh (bindings: 53)
Expected hash CoWBQoY2SykuFASd9UoVtQAWkH4jEHN9tMY2TJf8F9LHgcPUPeL8 but got CoVH63YNi2RdCV3Crzre2FCLHrnXTJC6EioHdhC7MXZddT5XL4bJ (bindings: 113)
Expected hash CoWF77dQ1ApRy6eqFy86TY2CKn3MWSXbsd27fkzzMVpuugaVzwHx but got CoUrYyEQNocD8Abd9ZXH8Cts4LYCCdaGYw9r1DYnYSxYuUubmNfB (bindings: 9)
Expected hash CoW6daoHhvYN8wmv2eLUS23W8RNNNoJ5pwYtjp7DkXW34cBg9cwS but got CoUootFjP2y4k62sGRHxtyMnnPS54Je5vtidVYBAXqjSnJH5mkvx (bindings: 44)
Expected hash CoVRUBPYsatU5TPTmvSEEWWV4u3MBWmZYv5JaW8jdRW42XvTyiFv but got CoWM42P8gXe9qFcYiL5aXXHgKZnFLnVmht9wP63guyCzWjPSK7M8 (bindings: 216)
Expected hash CoVqPRmvPDEird13jGPPutAqnuLrZPekz5qFpGYp9U7VamXbBmtL but got CoUehVoupdngCFfuAE9WueLjpAzGqvD6XWPHyZ8uSCRfPgX8gDmt (bindings: 200)
Expected hash CoVDKwEJnn5xCSheECj2GW1W5vC4rrococVA42rCR7BAHLUYjTWF but got CoVbuak4ZTfkkdxsFJcugwjGEpcUfLqGDv4BsHdHYnYHfdHn2Dzw (bindings: 205)
Expected hash CoW4DXbfdCCg11QTg2Zj3epHdN1B7xqQzQxmQ2L7n3bCi7XeRVzW but got CoWBTyTWYgNFwkUrW5LhJTpX1T4kodZW3pqzLMVG6YY7Qjq47bWw (bindings: 241)
Expected hash CoUx2abY1pVZKKJi8bQUNPU7U5jMXDi2ekh6w2DrVMuEunffG6pv but got CoVG5qRr2Dedq96omV2vqkuZA6dohLo4Qc9qscUYwsTiE1pzdPqF (bindings: 200)
Expected hash CoVtvftuE5Yu68WMahZNXocirYpa6sKbgVWzhYTvwnn38FK6Hhox but got CoVVJQkKsRPCAQh7PQH44QTDDQVQMF74WG4RKwmKHFt9Ahp3LaQL (bindings: 13)
Expected hash CoWaX3gGSoZSEem4Ei36XMLQea42A2xpmE9LwHp5brQUKB47Tvy9 but got CoVrLEHH3pUoQKieFCwKYjoVJmMMiVQQCi7MBfLgGYMYFsG9t4CC (bindings: 192)
Expected hash CoVrN1eTczySPJwkLZZwvqySEffTAzwa9qWDUFivKXKQe3FZvjux but got CoV7bzAKWscz6F1nkpjBvLS7zHLS2Yo1fjsL1vjnqy5ySgGExpMc (bindings: 238)
Expected hash CoUjWrf9NEMe8WJ8QW2K5FbbYq5CWrrRwj9z75NeG4edmy7QGR1q but got CoVeJ51xa7qKrQhq1TrbFLhK6kRbxAokVw1uABroizmFtmyRScMs (bindings: 55)
Expected hash CoWC9tnsVGkw2kLHS5dDDX6iuAVSqtPrCx2kCEMzPem5Ga6fJU8h but got CoVuUpu7ioq2s4XVtWraq6rVzNCLwvBeXstLRjPi2q1fcggCf186 (bindings: 178)
Expected hash CoWXa6FtAbwbhQr6AAWdfyLjYCGVapig5nvxqBtSxqUNfi4JhzYi but got CoVGMBExynLrwsJjUaAov5cZxrzmifmAXgcaDe1pMGU2rak2NbL8 (bindings: 57)
Expected hash CoVh4kGcWMJ65tFxi9KZVm1yAzzvBZQ71V77rQzn8Nc5QwGtWXFG but got CoWJBzzohzxBMrcsc4SwVS2uTLpp6LMHTn5DPN9khgDGrs7FPsBL (bindings: 89)
Expected hash CoV2z2KuJxWEeyFymG6jy1QqwVd2tiwJzDNER52hC36741RtuU7z but got CoVxnrb8wLbW7C88v5ArWDdTPRajyk581bJmgV4gdxKqsRZrNKB3 (bindings: 186)
Expected hash CoVUGD1reajD8DbFzphz5x11n6fTmLHS3yQq9QaHQpfDi4pUjQEs but got CoVm31aXciQvGiGu8c4fG2gf665WG33q6rXRD5Tr5uuq8qMZaiSh (bindings: 110)
Expected hash CoWSTNefVvBGGLWYu8Z4qU3jwHWNLPYTHV3jg7BiV6nNTPjKzRQZ but got CoVxK38ryayUNVHdxJihqtxvcBysroDA9E4uDujMGbGqFTEwvRw5 (bindings: 64)
Expected hash CoWR2yFJvvnUe7wmFYHHaDiUqZR975LzGnQRXPJ4bZ4c9coMw4MC but got CoV7cjYd2RPCPwc7ba3PKUM5rdRaQAytZrjaavsM3stpoJT6Dhcu (bindings: 89)
Expected hash CoV2eB5HKhwzbzc2TMpbJek4yuLBgTauSuFx3ZAnkQWj68CRNUu1 but got CoUvEC9mSacPQWAmA19NCB3JhHMsc4xceWTpmKQqViS3c1gkUzhY (bindings: 68)
Expected hash CoVGf593c4JXrb1MyHNvdGfbscGhmQMr33rGXmvLVJQgYBhVyo28 but got CoUr6PgJZN6qDCt8J9i5CDrrGdBc259JCBemJFbeVvWGQR8K9VqY (bindings: 210)
Expected hash CoWHoRiQbkHg2Ru6GQRWdoJqB7Rj6z5aJbe3v64J7o8qLQe22ed1 but got CoUfzBXAxfy4MGs7kF76ADDuyNpDhA6XsQzdeggHs7kTsgwB6dmL (bindings: 241)
Expected hash CoVNeqczQmo1Bb7Ed7F8reKAVZbwc9JCz48L1GedsVKCBPMobdar but got CoWLYZyG6sEjzUWtpFJ4Yur6aw53MaK1dNMk9QtzeSKSA64XJFd4 (bindings: 97)
Expected hash CoUtsjetNrbt7gQYj66ftNLTceiiTfRJHfwzeFwX9KvKCno8ZeLp but got CoVbxiffc8DcqLW9X79pwT2RdZk5Y1CWkAPDxLiNxS8xxEuxZddh (bindings: 92)
Expected hash CoUjbH129pjYrshuamDBGffzi3krV2PzijfGCxS8L5t5pTmDnFuJ but got CoV4wcgFji3kZGETDT5rr5CjHYWxAXe4CAPoQSPmTwoz9JkAASDu (bindings: 40)
Expected hash CoVxpLJcKZ63qL7NvkGq8yiWF7qpsxktCAnuRfsJ1URyqZeME4ZE but got CoUojTXMwn6a1MSqKusGN781PHsBqDUrMPxVgzKTrBzDXe2MctJG (bindings: 109)
Expected hash CoWGXQBU6jw4jz9ahVnPoxgkQACAP4TdDcFq3S8aLv8vDSfzpWCn but got CoW8ABNAvcn53kKGNEwMHB9QmwrUosr3WHP5K2D67vYZNGoErx82 (bindings: 182)
Expected hash CoWCs9uoap8AA2xSpRSrajsKNWJBCvaDGMpXazD3mjeaPYRXuL1q but got CoVj7DZZzQ9BKcFdW4obyDhfGAcmkphsRStZr1eDVxnr39d8bVdh (bindings: 66)
Expected hash CoWFacusA5iY6LdHaxAFzLigFpKjbrM7JXBrncpJmqHMAbskMBtP but got CoVWG1gVdW68p59eRWb9BUFq7ECnx2hYG44s4ZvHWT2wBRkrYZCT (bindings: 31)
Expected hash CoWXtj5dJ21ZeHUSVKPSQrnHtafRPSwhVCtMBEQm7gZwgFR2k3E5 but got CoVZk7YtDF6qapjLSL1W5RHRdiM7P4XqXKwVgQwD4PU1YGt1uHuq (bindings: 247)
Expected hash CoV8SMCMSqR4XUqEgFH5iqg61jXfo1aL7xzwAnRZdDqcDbtReZw2 but got CoVyMRSPqcmowJrv7L1k6dMNr9wHWkCK1r5GxeGdFsU9J5VfPfQF (bindings: 174)
Expected hash CoUo2AL6VA88E3NJWrqGA9mykUAo2n5paXS7mdeaXSGvw1d1iyRF but got CoVnqzy1RKNV4N8ycphKZrArUQwu4H55qi2Wob3KweAGpZC2Z7Uh (bindings: 25)
Expected hash CoVzWjnbXjLDhyv5tapiWx8qXXVC2RD9MRgRHWHJrQz7GhjxQrku but got CoUrHQWs28nh6mkzUrsj6EcnTFWTj35yVU5djFTkKnqbpq2Cjgxo (bindings: 118)
Expected hash CoWJHnkwCuJ1xhkY5vXsoLK8Hi9SceSd1LcV834EwKCesYZtrB5u but got CoUxrDW8aGJqAA3DC5qRhsNmPTMV3RRDtHt1tQQEnBT9UwwsspEr (bindings: 99)
Expected hash CoUth47DR1WRwn855mvMB6iGPpxbvwv6LLWUDQabR1hiy4pzrkrP but got CoW3TRhwvt1Mkn2jxTUJpkiMEakxJbSBYNNhQKYZbD98oJr4uZQe (bindings: 106)
Expected hash CoVyALmWrhKo4nXxZBH7JmhNjW4EP8iuUCAhqk2QXGKgA9Ea7Gz2 but got CoVnDM36LBPLXYNZ4hMKWF1P7tusLhTEr4AL3nGThYdeD5Mqp32C (bindings: 194)
Expected hash CoVUmKKqC9uz2RCPxpeyUgcMvCZ3kj2FQ48yoX5SiNStn4cHe3ny but got CoUw4h53mCQSjZgPHgBegLat919AZoLhFjdPNeaoXM2N16BPLdEi (bindings: 23)
Expected hash CoVVbTz769mw4PSHZJvynq7D5wkj4kqCS9imqezQJQ38AkmSgTbE but got CoW6o943NLpEf3usBYEAJ7Wc673XsSZxQfAhV2FEyVx4QHskufHj (bindings: 129)
Expected hash CoVqwbQWHy66tt5N2N6qAUKejL4dmtZhzWdht6uKchkdCxDc3YJN but got CoVZniHtHvmTYL4YGjRQqs5EZSna7ZyPYTe2umYyR7PPnuZ8scRo (bindings: 98)
Expected hash CoVBFcXzzropBdhKW3XPdGDYxt5xEyJrxizDLZBmnDGBRak756iw but got CoUknszvtb4v9QjrMB1coM1tgt1uVYJhJXSaEz9RfHSsStEg3VYK (bindings: 56)
Expected hash CoVPSp8eJuDX5BkUpoBTWZYrgMM5YPV6i1VpRWaXLcDYyoMUeseq but got CoWMRpjfV2yyuvTUXrVFARdGTgcmbzabrN1jy4GNbg41KAWcScXP (bindings: 109)
Expected hash CoVpHA4PA3m6WvMnXYT5GEj57EZwRsNkovNHGvjLcFtk5Md4Y7XV but got CoVq92DAEzJh4v3tEihkBgk4sfehVWn6XJtChDyfYVuRLayG1N4R (bindings: 37)
Expected hash CoVYM1Zm6CqYetUGomMVobdGwCwZ7HJS1q4NMB8sbVyR7T1Acswd but got CoVPiAqPaXUuKAr8TB5mp1gCUhx9FUYhrSbQixX4yd13h6sFttUn (bindings: 161)
Expected hash CoVmM4fd5iRBEHRmvkz3ymYKSmFPy75P4uYkHi7dMDmm24vJKgBr but got CoUzVNTKtPXDCxwAPMoS7GfdUaRYBBhCwqndKFZoEPNj6iJikJrT (bindings: 37)
Expected hash CoUvE55NgRdvfk7FSJyUDM1ri4GY7xvdNJpASM81Ak9ooWVJMQGY but got CoV6PzwidAkWFbqZmVYf3gSdawnVzm3SH7baQE7KnnFVsxZTRDpz (bindings: 118)
Expected hash CoW7ZQY69NDytRTLuBrCmAtGHQYbVJxX125LPMudDNEs5DpgaU3R but got CoV1D43EzGYSkqiyizYQKd4RrJkPdnbY7zEkN3GhUGDsDxCQaKeV (bindings: 25)
Expected hash CoVQQrzNqvvvDao5pQuFQEpgCzmXTDZrZjnHsPUH9w2PUHhRTGCA but got CoVYYwxSE2xQfRDoWr6Rcm9qY4jB8JxxouHk8VtzwwUMJkX3ixS6 (bindings: 5)
Expected hash CoVCnyZ6ZJeSPmirfEjzaSuViUobBGE9UGn5KmcAiCxKBZU6wG6T but got CoVA7Ryj82ZwBT6HYwp7Xwsn3inoMv4rAiGVVj482TwBNUjhtBMQ (bindings: 165)
Expected hash CoUvZJav3GeiArKSEpGPgwmzsXhk6su3GeNjT6XwcpqmM7VqfBRP but got CoW7ynxz5cWde3LqocvCNrJsuwZP7SDE2zSaRP65spmk1Lo1GP8L (bindings: 90)
Expected hash CoUukkbFHehigJK5k9J9wQwdAzeVMtc49ZZpNmgMSPhTUdXmntvQ but got CoVnqricLei239ZZvNMKNugbPnmR4Rz8dBWzVr5Xzk6g2rYBfa8D (bindings: 223)
Expected hash CoWShs5SvSmQnmqR63CCv9EvZEwKqWRhkP4GVUfiH958cJPNvATV but got CoUkHPm5sDsnxjdJP4PsX3X1kgy9L8fYQNK8TCYfkXkfbUZULW5M (bindings: 171)
Expected hash CoWXyuLDurGtjZqxCaGiV1iiAV2bW8NH1Ju2LXmFrBzhCmqf9VYY but got CoVQDpBq9EBHJHoqxFFywVXCSn1WU3b9kozdQcuvk7b2AuN9Qnem (bindings: 247)
Expected hash CoVmYDqhSicpicAP9RqKZzKe6iuhWiNjBHB1dzRTYbhd1qPLTcSA but got CoVK5Lgno8Cqq2NarCY2gJAcHuKW8mJ1wYrDSHnHZhu4fKutxh66 (bindings: 78)
Expected hash CoV3CR7jsfzASMevCpxKbV8Z6AW3MozLg9CDTPr6MYPuAYemhPzU but got CoW6W2zLnCfHS9JFVeEHZ3bYHhPefHc9HWQ3rUW4kheCJtk2F4f1 (bindings: 139)
Expected hash CoVvbW3cKoCmtrK5M3REgLDp5GuSHKG9QFe8AvjZi7Qs7H3x18Q6 but got CoW35bVdF58FeGe75FoAodmarBM9Yi7BZJ2nh4eTYxTbr1chcaoe (bindings: 23)
Expected hash CoWHh78czoJo3PFcwvjZrQ7QWBcbfq5bkM8xoTeuQBSjwhXsmxDS but got CoUfsLNc3a615ZoWmXbeYq67FbBQ69W8YkTkMwd7sfnU9pVL8XBH (bindings: 6)
Expected hash CoWFYC1R4r71YccaZxNTVxJiYGFEiR8nK39SMXvfvv1E5WELECnT but got CoVKrHsavFhq5S7c2B2JMZN1Vm6g39qt1VApcmRVxFb6FLnFDqQH (bindings: 112)
Expected hash CoVeTPMiS6BkcdvdgRU75nvypTX7i2YQCy7Ye95FuJ34T2S7nbix but got CoVfeic9mnx2ogytTk6oUmUctfNesheX7Gztn4G5uGU3hABbXrfz (bindings: 24)
Expected hash CoW4xfWChdpBQH27KyfLgGNgRrXE1f9JtCsSeWmwyq1VbzfaC2uz but got CoW5GzATpnzHCAqUER9GcsjmLyCt5sEM45TtaAobPUVr2fCeznam (bindings: 235)
Expected hash CoVF2ZcuhZV4HUU1Pwbq1s2uvZUqWJhfdNDSrvdNeJyjA3h31NDP but got CoVLryxbBCEBUo1Nz4oQetX3i1GLtmiT7Vyo5CKT6DjXXvhiTuxv (bindings: 98)
Expected hash CoVtk2vnitTJBDkVGZ5eDjQp2BbkDTupGqakCBWYuQ7EWdXS6eEz but got CoUpMoJvDPtY5mMzV5birkQKxxUwnzhpbzGJD6fnTzAizC8fCQR2 (bindings: 19)
Expected hash CoWK8Z7gLm95bkCDdHPju87v1BxWAQ11jHEkEYWfVHLbgU4kKQrj but got CoWUkeZWyRkhfGjgBJaC1XTQHQax7irVBhPwzCtBhWLzAVVmMwc4 (bindings: 37)
Expected hash CoWKS1huWR1DgNSkUtP8tfEd2i8yqD2RjjAVodZ2G2wCDvKrUdKq but got CoVEr2G5H9JkxUb2yQKmoBfDv6Qwa8u5f5843wU6kcMTwfSR1YPm (bindings: 61)
Expected hash CoUunoGdQ4bWYB8Atg9dM1EidQAujz9uARZSYqyrVKwT9LGDbozA but got CoW4XA6r71LhVc7nq6SHLkJNgQPmNkfnfiEZCPq3PP4cDV6N9Mtz (bindings: 147)
Expected hash CoUwCZysYYcHnq8boWf8hTJoMAhpE6mEyEhnD4d4a8pEbqFe5ANQ but got CoUqrHVij5ZCSfy3Rr1UPN3UnZUrUHwAUzaPQd1vQL2xchN2r8Ax (bindings: 209)
Expected hash CoWEpjWMsK2kTVivQpyh2cJkdyTJjd6X54nPgJEfzQDSsMzgU6ZL but got CoW58uJYun3siN6YGWdy8BQjC6euyCxwJcGXKCa8AtGa9djYHk4L (bindings: 186)
Expected hash CoVRmxVh81j7JoFQgv53pHYq8GFECAUaqoptM1cXT3bR7ojx5h4g but got CoVUH3TP49bpRU7E61S1Td5VusTqe1Y3RvjEKnYP5kei3284nT7h (bindings: 229)
Expected hash CoVybe8Dkx9dw1e5t3CYoZQZGBGGrhugcF1BsPwp5bHcnYfXFgCv but got CoV7XpkCPnVbdK19RAVyFyNdmfKxdUUSqzYkks7T4QBKMM6tCbqy (bindings: 134)
Expected hash CoVKKQiL2G7PMa18VwiksobckMxzNJd2Jzi4heWN13cQW8TJXqF3 but got CoVLSf5XgbrTLptUvwNqTx8CpAJHFZrNRXtGdksR3SANkJDWzpdt (bindings: 70)
Expected hash CoWahNJsicKJHHYEM6YTUEjQvFiz3EKMNQrkzRmsfaUsZqQjfYeR but got CoUqAaEYPZt29hN99WcixttzUynJ2Cjg8cYiShRXPDJ9k5Bx3sM8 (bindings: 178)
Expected hash CoV4MkpGeLGdTZcqpBmSx4xtStwi11kFAVXC9jAF2Avnt3QpXsF6 but got CoVREyKZ7PK8KVXAg2s6iTeFKucAhHycb1gGk2HuB9Zy2itG73ab (bindings: 49)
Expected hash CoUv66xsZYag7McLB7YjTx5SAunfE9rVnt6c11bWAnVawoLsCajn but got CoV3urbdo47KCqXUMfwf1ut5V7d4GvJgGEoSsmP7ZLfTo5WQoQm4 (bindings: 186)
Expected hash CoUnAK9QzDz4hEf1awGs5N26TcsfJqBmQmCwhMVKEqDP9pgussdV but got CoVzecpZAZMt73K3o1tyySCvANTbEXY9XseSx5XfucbVxQ5vDVaS (bindings: 175)
Expected hash CoVG8UBAyJBUNEjNpPRJnG44ug7YAudPDg7R4ZNo2Ngb6W6yqg3j but got CoVU5CqCRx5eAnaqSgDsinAM8QBvNpZUQwezKNx21q64Q9L59aSy (bindings: 218)
```

File: bin/main.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply bin/main.ml:202]: I don't know why it worked for large nodes, but I have an explanation on why it didn't work on short nodes:

I suppose that `Irmin.Hash.Typed (Store.Hash) (Store.Private.Node.Val)` uses the raw `.Val` module from `include M` (see [encoding/context.ml](https://gitlab.com/tezos/tezos/-/blob/master/src/lib_storage/encoding/context.ml#L137)), so the pre_hash misses.

**Comment by tizoc** [comment]
Also, with just the change I mentioned in #8, I get this running the program from this repo:

```
expected CoWCVQQRxmndUyNvCRpSn36viaGwp8MEJarqT3Xmkju5Apmj3WeK, got CoV8YLjMcZvaMFwddecEaXvcFjeFsak89PFjmc5SsNftPGx24u4L
```

Which matches the hash from the first case that fails in the above comment.

**Comment by samoht** [comment]
I found the issue: we were not using the right `Node` module 🤦 

I've force-pushed and rebased.

**Comment by tizoc** [comment]
Out of curiosity, what was the problem with flambda?

File: bin/main.ml [hunk 2]
```diff
@@ -29,23 +27,16 @@ module Spec = struct
...
-    let bt = Inter.Val.to_bin t in
-    {
-      hash = Inter.Elt.hash bt;
-      bindings = List.map entry (Inter.Val.list t);
-      inode;
-    }
+    let stable = Inter.Val.stable t in
+    let inode = if stable then None else Some (inode t) in
+    let hash = I.hash t in
+    { hash; bindings = List.map entry (Inter.Val.list t); inode }
```
Comment (mattiasdrp) [inline bin/main.ml:36]: Is this a simplification or did it have an impact on the bug? On my test files I didn't make this change, just changed 

```ocaml
module N = Irmin.Hash.Typed (Store.Hash) (Inter.Val)
```

deleted the Node module and then I generated the dataset and checked and it worked.

File: bin/main.ml [hunk 3]
```diff
@@ -1,9 +1,7 @@
 open Irmin_tezos
 open Encoding
-module Node = Irmin.Private.Node.Make (Hash) (Path) (Metadata)
```
Comment (mattiasdrp) [inline bin/main.ml:3]: Good catch!

File: bin/main.ml (hunk 2 above)
Comment (samoht) [reply bin/main.ml:36]: yes it's a simplification - I'll revert it to make things clearer.

**Comment by samoht** [comment]
@tizoc: flamdba tries to inline side-effect functions which changed the order of evaluation of the various calls to `Random.*`. I don't think that's expected. https://github.com/tarides/tezos-context-hash/pull/9/commits/3297370659447f610b2b0f073d94c58f91797b39 ensure that the functions kept being evaluated in the same order.

@Ngoguey42 is creating a small repro case and will open an issue on ocaml/ocaml.

**Comment by samoht** [comment]
@mattiasdrp yes please feel free to open PRs to simplify the code :-)

**Pull request merged**